### PR TITLE
GH-93354: Use exponential backoff to avoid excessive specialization attempts.

### PR DESCRIPTION
**Pull request by markshannon**

### CODE REVIEW

**Comment by markshannon** [comment]
[No significant change in performance](https://gist.github.com/markshannon/c29e72db6e3b34530aa0638c6089e9e3) on the standard benchmarks, as expected.

File: Include/internal/pycore_code.h [hunk 1]
```diff
@@ -227,8 +227,10 @@ extern void _PyLineTable_InitAddressRange(
 extern int _PyLineTable_NextAddressRange(PyCodeAddressRange *range);
 extern int _PyLineTable_PreviousAddressRange(PyCodeAddressRange *range);
 
-
-#define ADAPTIVE_CACHE_BACKOFF 64
+/* With a 16 counter, we have 12 bits for the counter value, and 4 bits for the backoff */
+#define ADAPTIVE_BACKOFF_BITS 4
+/* The initial counter value is backoff**2-1, so a backoff of 5 gives a value of 31 */
```
Comment (sweeneyde) [inline Include/internal/pycore_code.h]:
```suggestion
/* The initial counter value is 2**backoff-1, so a backoff of 5 gives a value of 31 */
```

File: Include/internal/pycore_code.h [hunk 2]
```diff
@@ -227,8 +227,10 @@ extern void _PyLineTable_InitAddressRange(
 extern int _PyLineTable_NextAddressRange(PyCodeAddressRange *range);
 extern int _PyLineTable_PreviousAddressRange(PyCodeAddressRange *range);
 
-
-#define ADAPTIVE_CACHE_BACKOFF 64
+/* With a 16 counter, we have 12 bits for the counter value, and 4 bits for the backoff */
```
Comment (sweeneyde) [inline Include/internal/pycore_code.h]:
```suggestion
/* With a 16-bit counter, we have 12 bits for the counter value, and 4 bits for the backoff */
```

Comment (sweeneyde) [reply Include/internal/pycore_code.h]: Maybe a longer comment on the counter format would be useful as well? E.g. explaining that the LSBs `counter & 0b1111` are named "backoff" and represent an exponent, while the MSBs `counter >> 4` represent the number of executions until the next specialization attempt, initialized based on the backoff exponent.

**Pull request merged**